### PR TITLE
feat(react-ag-ui): attribute AG-UI subagent lifecycle into ToolCallMessagePart.messages

### PR DESCRIPTION
**Pull request by farazhassan**
## Problem

Closes (or at least gives a concrete starting point for) #6529. `packages/react-ag-ui` doesn't model `@ag-ui/client@0.0.59+`'s subagent protocol: `SUBAGENT_STARTED`/`SUBAGENT_FINISHED`/`SUBAGENT_ERROR` events and the `subagentRunId` attribution field carried on most other events. `event-parser.ts`'s `default` branch folds all three into a `RAW` event, and `run-aggregator.ts` logs and drops `RAW`. A backend that runs subagents (the "agents-as-tools" pattern, e.g. deepagents' `task` tool) produces output with no lifecycle boundaries and no attribution — a subagent's own tool calls and text render as flat siblings of the parent run, with no way to know which activity belongs to which subagent.

## Root cause

The adapter's event model (`AgUiEvent` in `types.ts`) and its aggregator (`RunAggregator`) predate the subagent protocol addition and never had a concept of "this event belongs to a nested agent run" — every accumulator (`textParts`, `reasoningParts`, `toolCalls`, `partOrder`) was single-scope by construction, and events with an unmodeled `type` fall through to a defensive `RAW` catch-all that's dropped at debug level.

## Change

**Protocol facts** (confirmed against a fresh clone of `ag-ui-protocol/ag-ui`, not guessed or taken from minified dist — the issue posed the correlation mechanism as an open question, so this is worth stating precisely):
- `SUBAGENT_STARTED` carries `subagentRunId`, `name`, `description?`, `parentSubagentRunId?`, and **`parentToolCallId?`/`parentMessageId?`** — the correlation back to the spawning tool call is explicit on the event itself, not inferred from ordering. `parentToolCallId` is what this PR's nesting join actually keys on; `parentSubagentRunId` is stored and surfaced as descriptive metadata on the synthesized nested message but doesn't drive the join itself (nesting depth is driven entirely by `parentToolCallId` chains).
- `SUBAGENT_FINISHED` has `outcome?: {type:"success"} | {type:"suspended", interruptIds?}`, mirroring `RUN_FINISHED.outcome` one level down. `SUBAGENT_ERROR` has `message`/`code?`.
- `subagentRunId?` is on nearly every event schema; `@ag-ui/client`'s own reducer only tags it flatly onto the `Message`/`ToolCall` objects it builds — it does **not** nest or group anything. Grouping is left to the consumer, by design.
- Cross-runtime precedent: `@assistant-ui/react-langchain`'s current (v1-era) subagent support is discovery-only (`useLangChainSubagents()` returns a namespace-keyed map; fetching the actual messages is left to the caller via LangChain's own `useMessages`). The **earlier** legacy-runtime pattern (assistant-ui#2574/#2580/#2583) is the one that actually built and attached `ToolCallMessagePart.messages` — whoever converts a protocol's events into `ThreadMessage[]` is responsible for using that protocol's own correlating concept and assigning the array directly. AG-UI's explicit `parentToolCallId` makes this more precise than LangChain's namespace scheme, so this PR follows that established core contract rather than inventing a parallel mechanism — and doesn't add a discovery-hook API, since the nested `.messages` tree already is the discovery mechanism here.

**Implementation**, scoped to `packages/react-ag-ui`, no `@assistant-ui/core` changes needed (`ToolCallMessagePart.messages` and its consumers — `PartPrimitive.Messages`, `ToolInvocationTracker`, `external-store-thread-runtime-core`'s identity walk — already handle nested `ThreadMessage[]` generically):

1. `types.ts` — `subagentRunId?: string` added to the `AgUiEvent` variants that need it for attribution (text/reasoning/tool-call events, including `REASONING_ENCRYPTED_VALUE`); three new variants for the lifecycle events.
2. `event-parser.ts` — parses all of the above instead of falling through to `RAW`.
3. `subscriber.ts` — three new typed `AgentSubscriber` callbacks (matching `@ag-ui/client`'s real `onSubagentStartedEvent`/`onSubagentFinishedEvent`/`onSubagentErrorEvent` names).
4. `run-aggregator.ts` — the core of the change. `RunAggregator`'s existing flat accumulators are reused as-is (ids are unique per run regardless of subagent depth, per the protocol schema), but every entry is tagged with an optional `subagentRunId` scope as it's created, and a `subagentRuns` registry tracks each run's lifecycle/metadata. `emit()`'s snapshot construction became a recursive `buildParts(scope, depth, ...)`: root-scope parts assemble as before; any tool-call part whose `toolCallId` matches a subagent's `parentToolCallId` gets that subagent's own parts (rebuilt live on every emit, so nesting streams in incrementally rather than only appearing after `SUBAGENT_FINISHED`) materialized into a synthetic `ThreadAssistantMessage` and attached via `.messages` — recursing for subagent-of-subagent chains, with a depth cap as a defensive (not expected-case) guard against malformed cycles.

Two correctness issues surfaced and fixed along the way, both from a previously-single-scope collection quietly becoming multi-scoped:
- The aggregator previously tracked "which buffer should an anonymous (no explicit `messageId`) delta append to" via two single scalar fields. Once subagent-tagged events can interleave with the top-level run (or with each other, if a backend runs subagents concurrently), a shared scalar would misattribute anonymous deltas across scopes. Replaced with `Map<scope, ...>` keyed by `subagentRunId ?? ""`.
- Tool-approval projection (`projectAgUiToolApprovals`'s `boundToolCallIds` set, in the pre-existing `tool-approval.ts` seam added in #5902) was being passed `this.toolCalls.keys()` unfiltered by scope. Once subagent-scoped tool calls exist, a `RUN_FINISHED` interrupt batch mixing a root-level approval gate with a subagent-scoped one would get miscategorized as "fully bound" by the projection call, when the subagent's gate actually has no UI anywhere to answer it — and `buildToolApprovalResume`'s existing all-or-nothing contract means the run could never resume once a mismatched approval keeps a batch open. Fixed by scoping `boundToolCallIds` to root-only tool calls, which correctly lets `tool-approval.ts`'s existing "a batch containing an unbound id projects nothing" rule apply to the whole mixed batch (matching its pre-existing documented contract) rather than silently misrendering it as actionable.

## Verification

- `pnpm --filter @assistant-ui/react-ag-ui test` — 498/498 passing (all pre-existing tests green with zero regressions, plus new coverage: subagent event parsing incl. edge cases and `REASONING_ENCRYPTED_VALUE` attribution, subscriber dispatch, scoped anonymous-delta attribution, single-level nesting, two-level subagent-of-subagent recursion, `SUBAGENT_ERROR` preserving partial accumulated content instead of dropping it, duplicate-`parentToolCallId` ordering, an orphaned-subagent (unmatched `parentToolCallId`) non-crash case, mixed-scope approval-batch collapsing correctly, a dangling subagent-scoped tool call reporting `incomplete`/`tool-calls` rather than `requires-action` or a silent `complete`, a root-scoped unresolved tool call still correctly reporting `requires-action`/`tool-calls` alongside an already-resolved subagent-scoped one, and one true end-to-end test driving raw wire-shaped JSON through `createAgUiSubscriber` → `parseAgUiEvent` → `RunAggregator` → nested `.messages`).
- `pnpm --filter @assistant-ui/react-ag-ui build` — clean.
- `pnpm lint` — clean (repo-wide warnings present are pre-existing and unrelated to this change).

## Public surface

- `@assistant-ui/react-ag-ui`: additive only. `AgUiEvent` gains `subagentRunId?` on several variants and three new discriminated variants (`SUBAGENT_STARTED`/`FINISHED`/`ERROR`); `Subscriber` gains three new optional callbacks. `ToolCallMessagePart.messages` (existing `@assistant-ui/core` field) is now populated by this adapter for the first time — no type change there.
- Changeset added (patch).
- No documentation/api-reference changes — nothing here is part of the generated API reference surface (internal adapter types), and I didn't see an existing docs page describing react-ag-ui's event coverage to update; happy to add one if maintainers want it.

## Explicitly deferred / open questions for maintainers

Per the issue's own framing ("requires maintainer decisions", "not scoped for automation"), I'm not presenting this as a finished, settled design — flagging what I deliberately left out or am least sure about:

- **Scope, stated explicitly rather than silently:** one `ThreadMessage` per subagent run (not per distinct wire `messageId` within it — a subagent's entire output is one nested message, even if the subagent itself emits multiple `TEXT_MESSAGE_START` boundaries). `CUSTOM`/`STATE_SNAPSHOT`/`STATE_DELTA` don't carry attribution in this PR even though the protocol schema has `subagentRunId` on all of them; the scope is what the text/reasoning/tool-call/reasoning-signature attribution path needs.

  **Corrected in a follow-up commit on this branch.** An earlier revision of this section said `ACTIVITY_SNAPSHOT` "carries no `subagentRunId` at all" and left MCP-apps routing root-scope-only on that basis. That was wrong: `ActivitySnapshotEventSchema` and `ActivityDeltaEventSchema` both declare `subagentRunId` in `@ag-ui/core@0.0.59`, and upstream's own reducer (`apply/default.ts`) implements ownership semantics for it. `ACTIVITY_SNAPSHOT`, `TEXT_MESSAGE_CHUNK` and `TOOL_CALL_CHUNK` now parse the field, and the MCP-apps fallback resolves through a per-scope pointer.
- **`SUBAGENT_FINISHED`'s `result` and a suspended outcome's `interruptIds`** are now kept on the synthesized nested message's `agui` metadata rather than discarded, but nothing consumes them yet.
- **`SUBAGENT_FINISHED`'s `suspended` outcome** maps onto `MessageStatus`'s existing `requires-action`/`interrupt` shape, but I haven't built any nested-HITL resume UX on top of it — happy to hear whether that's wanted here or is a separate follow-up.
- **No discovery-hook API** (mirroring react-langchain's `useLangChainSubagents()`) was added — reasoning above is that the nested `.messages` structure already answers "does this tool call have subagent activity," but if there's a UX reason to want pre-completion discoverability before any content streams in, that's an easy addition on top of the existing `subagentRuns` registry.
- **A subagent with no nesting site flattens into the root scope** rather than being dropped. `parentToolCallId` is optional in the schema, so dropping silently discarded every token from a producer that never sets it; flattening is also what the protocol's own pre-subagent compatibility middleware does.
- **Known, not fixed in this PR:** `AgUiThreadRuntimeCore.findMessageIdForToolCall` only scans top-level `message.content`, not nested `.messages` — a subagent's own tool call resolving asynchronously across a run boundary (the cross-run `TOOL_CALL_RESULT`/MCP-apps `ACTIVITY_SNAPSHOT` paths) won't find it in a historical nested subagent message. Likely rare in practice; flagging as a probable follow-up rather than expanding this PR's surface further.
- **Subagent-scoped client-side (frontend-executed) tool calls are not resolvable by this adapter.** The protocol allows a subagent to call a tool from `RunAgentInput.tools`, the same as the root run — nothing in the schema restricts frontend tools to the root agent. But `AgUiThreadRuntimeCore.getPendingToolCalls()`/`findMessageIdForToolCall()` only walk a message's top-level `content`, not nested `.messages`, so a subagent's own unresolved tool call can never be surfaced to `addToolResult`/resumed — reaching it would mean extending the pending-tool-call and resume paths to walk the nested tree, which is a real feature in its own right and out of scope here. Given that ceiling, `RUN_FINISHED` deliberately does **not** report `requires-action`/`tool-calls` for a run whose only unresolved tool calls are subagent-scoped (there would be no way for a caller to act on it), but it also does not report `complete` (that would silently tell the app the run finished successfully while an AG-UI backend may still be blocked waiting on a `TOOL_CALL_RESULT` that will never come). It reports `{ type: "incomplete", reason: "tool-calls" }` instead — the same status `@assistant-ui/core`'s local runtime already uses when it deliberately stops resolving tool calls (`local-thread-runtime-core.ts`'s `maxSteps` cutoff) — so the failure is visible and honestly labeled rather than either silently swallowed or falsely reported as resumable. Full nested tool-call execution support is a natural follow-up if subagent-owned frontend tools turn out to be a real pattern maintainers want supported.

Open to feedback on any of the above, or on the overall "one nested message per subagent run" shape if there's a different design maintainers would prefer.

### CODE REVIEW

File: packages/react-ag-ui/src/runtime/adapter/run-aggregator.ts [hunk 1]
```diff
@@ -167,7 +188,7 @@ export class RunAggregator {
 
         this.interrupts = undefined;
         const hasUnresolvedToolCalls = Array.from(this.toolCalls.values()).some(
-          (tc) => tc.result === undefined,
+          (tc) => tc.subagentRunId === undefined && tc.result === undefined,
```
[the remark this answers is not in the document]
Comment (farazhassan) [reply packages/react-ag-ui/src/runtime/adapter/run-aggregator.ts]: Good catch, and you're right that neither original option was acceptable — fixed in 9ef5dcae1.

Confirmed the mechanism: `AgUiThreadRuntimeCore.getPendingToolCalls()` and `findMessageIdForToolCall()` only walk a message's top-level `content`, not nested `.messages`, so a subagent-scoped tool call genuinely cannot be surfaced for execution/resume through this adapter today — reaching it would mean extending the pending-tool-call and resume paths to walk the nested tree, which is a real feature in its own right and beyond this PR's scope (subagent lifecycle attribution, not subagent tool execution).

Given that, `requires-action`/`tool-calls` would be a lie (nothing can ever answer it), so I did not revert to that. Instead `RUN_FINISHED` now reports `{ type: "incomplete", reason: "tool-calls" }` when only subagent-scoped tool calls are unresolved — the same status `@assistant-ui/core`'s local runtime already uses when it deliberately stops resolving tool calls (`local-thread-runtime-core.ts`'s `maxSteps` cutoff), so this is an established, correctly-typed status for exactly "there are tool calls this runtime is not going to resolve," rather than a status this PR invents. That keeps the previously-passing root-only cases unchanged (a root-scoped unresolved tool call still correctly forces `requires-action`/`tool-calls`, now with a test proving that alongside an already-resolved subagent-scoped call) while making the failure honest instead of silent.

Also wrote this up explicitly in the PR body's "Explicitly deferred" section, called out as its own bullet, per your ask — not just in a test.

498/498 tests passing (added a positive-status test alongside the corrected dangling-tool-call test), build and lint clean.

File: packages/react-ag-ui/src/runtime/types.ts [hunk 2]
```diff
@@ -136,34 +140,58 @@ export type AgUiEvent =
...
   | { type: "RUN_ERROR"; message?: string; code?: string }
-  | { type: "TEXT_MESSAGE_START"; messageId?: string }
-  | { type: "TEXT_MESSAGE_CONTENT"; messageId?: string; delta: string }
-  | { type: "TEXT_MESSAGE_END"; messageId?: string }
+  | { type: "TEXT_MESSAGE_START"; messageId?: string; subagentRunId?: string }
+  | {
+      type: "TEXT_MESSAGE_CONTENT";
+      messageId?: string;
+      delta: string;
+      subagentRunId?: string;
```
[the remark this answers is not in the document]
Comment (okisdev) [reply packages/react-ag-ui/src/runtime/types.ts:148]: already there. `types.ts` declares `subagentRunId?: string` on the `REASONING_ENCRYPTED_VALUE` variant, `event-parser.ts` reads it, and the aggregator routes the visible path through `this.scopeOf(event)` before it looks up the block to claim. what stays root-only is the hidden-reasoning bookkeeping (the `showThinking === false` path), which is a stated scope limit rather than a missing field.

File: packages/react-ag-ui/src/runtime/adapter/run-aggregator.ts [hunk 3]
```diff
@@ -390,6 +433,52 @@ export class RunAggregator {
...
+          createdAt: new Date(),
+          status: { type: "running" },
+        });
+        this.emit();
+        break;
+      }
+      case "SUBAGENT_FINISHED": {
+        const run = this.subagentRuns.get(event.subagentRunId);
+        if (run) {
+          run.status =
```
[the remark this answers is not in the document]
Comment (okisdev) [reply packages/react-ag-ui/src/runtime/adapter/run-aggregator.ts:532]: half accepted. `interruptIds`, `SUBAGENT_FINISHED.result` and `SUBAGENT_ERROR.code` are now kept on the nested message's `agui` metadata instead of being discarded.

keeping `requires-action`/`interrupt` on the nested message is deliberate. that status describes the subagent run; the run-level status the resume APIs read is computed separately, from the root scope, so nothing here surfaces an unanswerable gate through `getPendingToolCalls`. reporting `complete` for a subagent that is actually parked would be the worse lie. nested HITL resume stays out of scope, and preserving the ids is what makes it buildable later rather than a re-derivation from a later interrupt event.

File: packages/react-ag-ui/src/runtime/adapter/run-aggregator.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (okisdev) [reply packages/react-ag-ui/src/runtime/adapter/run-aggregator.ts]: the run has never reported `complete` for a dangling subagent-scoped call on this branch; it reports `{ type: "incomplete", reason: "tool-calls" }`.

the classification is now sharper than that. it keys on the scope a call actually renders in: a subagent with no nesting site flattens to root, so its call is top-level content, `getPendingToolCalls` reaches it, and the run reports `requires-action`/`tool-calls` and executes the tool normally. only a call nested inside a subagent message stays `incomplete`, and that one genuinely has no resume path in this adapter. both directions are covered by tests.

File: packages/react-ag-ui/test/run-aggregator.spec.ts [hunk 4]
```diff
@@ -2824,4 +2937,886 @@ describe("RunAggregator", () => {
...
+      type: "text",
+      text: "checking dashboards",
+    });
+    const nestedToolPart = nested.content[1] as any;
+    expect(nestedToolPart.type).toBe("tool-call");
+    expect(nestedToolPart.toolCallId).toBe("t-explore");
+    expect(nestedToolPart.result).toEqual({ p99: 420 });
+  });
+
+  it("resolves nested subagent tool-call messages from raw wire-shaped JSON, through the subscriber's dispatch, not just hand-typed AgUiEvent objects", () => {
```
[the remark this answers is not in the document]
Comment (okisdev) [reply packages/react-ag-ui/test/run-aggregator.spec.ts:3479]: accepted as a gap, not fixed here. nested tool execution is the deferred surface: `getPendingToolCalls` and `findMessageIdForToolCall` walk top-level content only, so a call nested inside a subagent message has no resume path by construction, and a core-level test would be pinning that absence rather than a behaviour.

the reachable path is covered at the seam where the decision is actually made: a subagent with no nesting site flattens to root and the run reports `requires-action`/`tool-calls`, a nested one stays `incomplete`. extending the pending-tool and resume walks into `.messages` is the follow-up that would make a core-level test meaningful.

File: packages/react-ag-ui/src/runtime/adapter/run-aggregator.ts [hunk 5]
```diff
@@ -754,54 +1052,86 @@ export class RunAggregator {
...
+          : ROOT_SCOPE;
+      const bucket = partsByScope.get(bucketScope);
+      if (bucket) bucket.push({ index, part });
+      else partsByScope.set(bucketScope, [{ index, part }]);
     }
+    const ctx: BuildContext = {
+      subagentsByParentToolCallId,
+      reachable,
+      partsByScope,
+      approvals: projectAgUiToolApprovals(
```
[the remark this answers is not in the document]
Comment (okisdev) [reply packages/react-ag-ui/src/runtime/adapter/run-aggregator.ts:1161]: deliberate, and it follows the projector's existing contract rather than a new choice. `projectAgUiToolApprovals` collapses the whole batch when any gate names an unbound call, because a partially rendered batch can never be completed: the user answers the half that rendered while the resume array still waits on an id no part carries. a nested call is unbound precisely because `getPendingToolCalls` cannot reach it.

so the options are a nested tool call with no approval surface, or a root gate the user can click that leaves the run stuck. the first is the honest one while nested HITL is unsupported. `SUBAGENT_FINISHED.outcome.interruptIds` are now preserved on the nested message, which is what a later nested-resume implementation would key on.

File: packages/react-ag-ui/src/runtime/adapter/run-aggregator.ts [hunk 6]
```diff
@@ -166,13 +209,29 @@ export class RunAggregator {
...
+            tc.subagentRunId !== undefined && reachable.has(tc.subagentRunId),
+        );
 
-        this.status = hasUnresolvedToolCalls
+        this.status = hasUnresolvedRootToolCalls
           ? { type: "requires-action", reason: "tool-calls" }
-          : { type: "complete", reason: "unknown" };
+          : hasUnresolvedSubagentToolCalls
+            ? { type: "incomplete", reason: "tool-calls" }
+            : { type: "complete", reason: "unknown" };
```
[the remark this answers is not in the document]
Comment (okisdev) [reply packages/react-ag-ui/src/runtime/adapter/run-aggregator.ts:241]: confirmed, and it is the one finding on this branch i am not going to paper over. before this PR `subagentRunId` was unparsed, so a subagent's frontend tool call sat in top-level `content` and the whole execute/`addToolResult`/resume path worked by accident. once the call nests, it does not.

i looked at what closing it takes. `getPendingToolCalls` and `findMessageIdForToolCall` walking `.messages` is easy. the rest is not: `addToolResult` maps top-level content only, `preserveToolResults` is what stops the aggregator clobbering an injected result and also maps top-level only, and `buildParts` regenerates every nested message from `subagentRuns` plus `toolCalls` on each emit, so a result written into a nested `ThreadMessage` is discarded on the next delta unless the aggregator's own tool-call entry holds it. that is the "real feature in its own right" the description already calls out, not a follow-up commit on this one.

so this thread stays open. the run reporting `incomplete`/`tool-calls` keeps the failure visible rather than silent, which is the right interim, but it is a capability that used to work.

[the remark this answers is not in the document]
Comment (okisdev) [reply packages/react-ag-ui/src/runtime/adapter/run-aggregator.ts:241]: resolved by writing it down rather than by changing the code, which is the honest disposition for a gap this PR is not going to close.

the subagent protocol shipped in `@ag-ui/core@0.0.59` on 2026-08-27, four days ago, so the window in which a backend could depend on subagent frontend tools working is that wide, and those same backends are currently getting `SUBAGENT_*` dropped as RAW with every subagent part flattened into undifferentiated siblings. trading that for nested-but-not-resumable is a net improvement, but it is still a capability that used to work.

so it is now on the record in three places instead of one review thread: the react-ag-ui README has a subagents section naming it, the changeset carries it as a behavior change so it reaches release notes, and #6612 has the full account of what a fix has to cover (the two lookups are the easy half; `addToolResult`, `preserveToolResults`, `cancelUnresolvedToolCalls` and the aggregator regenerating nested messages on every emit are the rest).

**Review by okisdev — APPROVED** [review]
approved, and i pushed the follow-ups onto your branch rather than sending them back, since nearly all of them fell out of one wrong fact rather than a wrong design.

the protocol work is the part i would have had to redo from scratch if it were wrong, and it isn't. i checked every claim in the description against `@ag-ui/core@0.0.59` at upstream `release/2026-08-27`: `SubagentStartedEventSchema` (`events.ts:401`), the client reducer's flat tag (`apply/default.ts:434`), the subscriber callback names (`:1412-1455`). they hold verbatim. picking `parentToolCallId` as the join and `ToolCallMessagePart.messages` as the target is right for this repo.

the claim that didn't hold was `ACTIVITY_SNAPSHOT` carrying no `subagentRunId`. it does, and upstream implements ownership semantics for it in `apply/default.ts:846-875`, down to a rule about a subagent not losing an activity it claimed. that one fact is what kept MCP-apps routing root-scope-only; with it fixed, a subagent's tool result no longer captures a root activity that lands after it. `TEXT_MESSAGE_CHUNK` and `TOOL_CALL_CHUNK` were the same shape, so a producer using the compact form was getting zero attribution.

the design change worth arguing over is dropping. your test asserted an orphaned subagent's output vanishes rather than flattening into root, and i inverted it. `parentToolCallId` is optional in the schema, so that rule silently discarded every token from any producer that never sets it, which is not an edge case but the default for anything not implementing agents-as-tools. upstream hit the same "cannot represent nesting" problem and chose flatten: `client/src/middleware/backward-compatibility-0-0-57.ts` strips `subagentRunId`, keeps the events, and documents it as "the intended downgrade, since subagent separation cannot be represented to a consumer with no concept of it". the status classification then falls out cleanly: a call that flattens to root is reachable by `getPendingToolCalls` and reports `requires-action`, a nested one is not and stays `incomplete`.

the rest was mechanical. `messageId` is not unique across subagent runs anywhere in the schema, so the text and reasoning accumulators key on scope too. `emit()` walked `partOrder` twice with the materialization predicate copied into both halves. a repeated `SUBAGENT_STARTED` re-parented an established run, which upstream describes as a resume continuation. a run terminal left a still-streaming subagent's nested message spinning on `running` forever. and a type error was already on the branch: routing the narrowed `subtype` of `REASONING_ENCRYPTED_VALUE` through the generic `withOptional` widened it back to `string`, with CI green over it.

what i executed, as opposed to read: `pnpm --filter @assistant-ui/react-ag-ui test` goes 498 to 510, and each of the nine new behaviour tests was confirmed to fail against the pre-fix tree before i kept it (`git stash push -- packages/react-ag-ui/src`, run, pop). `tsc --noEmit`, oxlint, oxfmt and `aui-build` are clean on the package. what i only read is the behaviour against a real AG-UI backend; none of this has been driven end to end against a live subagent producer, and the cheap check that would catch a wiring mistake there does not exist in this repo yet.

one thing stays open, and it is the honest cost of the change. you flagged it yourself: a subagent's frontend-executed tool call is unreachable once nested, and that path worked before only because `subagentRunId` was never parsed. i traced what closing it takes and agree it is not a commit on this PR. the two lookups are easy, but `addToolResult` and `preserveToolResults` both map top-level content only, and `buildParts` regenerates every nested message from the registry on each emit, so an injected result is discarded on the next delta unless the aggregator's own tool-call entry holds it. that is #6612, it is in the readme's subagents section, and it rides in the changeset as a behaviour change so it reaches release notes.

the trap for whoever reads this next: `parentToolCallId` is optional, so nesting is never guaranteed. anything assuming a subagent's parts live inside a nested message has to handle them appearing at root instead.
